src/Servers/Kestrel/Core/src/Internal/Infrastructure/Heartbeat.cs

```diff
@@ -62,26 +63,30 @@ internal void OnHeartbeat()
         }
         catch (Exception ex)
         {
-            if (!_stopped)
-            {
-                _trace.LogError(0, ex, $"{nameof(Heartbeat)}.{nameof(OnHeartbeat)}");
-            }
+            _trace.LogError(0, ex, $"{nameof(Heartbeat)}.{nameof(OnHeartbeat)}");
         }
     }
 
     private void TimerLoop()
     {
-        Thread.Sleep(_interval);
-        while (!_stopped)
+        // Starting the heartbeat immediately triggers OnHeartbeat.
+        // Initial delay to avoid running heartbeat again from timer thread.
+        while (!_stopWaitHandle.WaitOne(_interval))
         {
             OnHeartbeat();
-            Thread.Sleep(_interval);
         }
     }
 
     public void Dispose()
     {
-        _stopped = true;
-        // Don't block waiting for the thread to exit
+        // Stop heart beat and immediately exit wait interval.
+        _stopWaitHandle.Set();
+
+        // Wait for heartbeat thread to finish.
+        // Should either be immediate or a short delay while heartbeat callbacks complete.
+        if (_timerThread.IsAlive)
```

Comment:
total nit, I think IsAlive is redundant here.

Reply:
Some unit tests call dispose without calling start. `Thread.Join` breaks if the thread isn't started.